cpp/include/raft/spatial/knn/detail/topk/radix_topk.cuh

```diff
@@ -0,0 +1,562 @@
+  IdxT out_cnt;
+  IdxT out_back_cnt;
+};
+
+/**
+ * Fused filtering of the current phase and building histogram for the next phase
+ * (see steps 4-1 in `radix_kernel` description).
+ */
+template <typename T, typename IdxT, int BitsPerPass>
+__device__ void filter_and_histogram(const T* in_buf,
+                                     const IdxT* in_idx_buf,
+                                     T* out_buf,
+                                     IdxT* out_idx_buf,
+                                     T* out,
+                                     IdxT* out_idx,
+                                     IdxT len,
+                                     Counter<T, IdxT>* counter,
+                                     IdxT* histogram,
+                                     bool greater,
+                                     int pass,
+                                     int k)
+{
+  constexpr int num_buckets = calc_num_buckets<BitsPerPass>();
+  __shared__ IdxT histogram_smem[num_buckets];
+  for (IdxT i = threadIdx.x; i < num_buckets; i += blockDim.x) {
+    histogram_smem[i] = 0;
+  }
+  __syncthreads();
+
+  const int start_bit = calc_start_bit<T, BitsPerPass>(pass);
+  const unsigned mask = calc_mask<T, BitsPerPass>(pass);
+
+  if (pass == 0) {
+    auto f = [greater, start_bit, mask](T value, IdxT) {
+      int bucket = calc_bucket<T, BitsPerPass>(value, start_bit, mask, greater);
+      atomicAdd(histogram_smem + bucket, IdxT(1));
+    };
+    vectorized_process(in_buf, len, f);
+  } else {
+    const IdxT previous_len      = counter->previous_len;
+    const int want_bucket        = counter->bucket;
+    IdxT& filter_cnt             = counter->filter_cnt;
+    IdxT& out_cnt                = counter->out_cnt;
+    const IdxT counter_len       = counter->len;
+    const int previous_start_bit = calc_start_bit<T, BitsPerPass>(pass - 1);
+    const unsigned previous_mask = calc_mask<T, BitsPerPass>(pass - 1);
+
+    auto f = [in_idx_buf,
+              out_buf,
+              out_idx_buf,
+              out,
+              out_idx,
+              greater,
+              k,
+              start_bit,
+              mask,
+              previous_start_bit,
+              previous_mask,
+              want_bucket,
+              &filter_cnt,
+              &out_cnt,
+              counter_len](T value, IdxT i) {
+      int prev_bucket =
+        calc_bucket<T, BitsPerPass>(value, previous_start_bit, previous_mask, greater);
+      if (prev_bucket == want_bucket) {
+        IdxT pos     = atomicAdd(&filter_cnt, IdxT(1));
+        out_buf[pos] = value;
+        if (out_idx_buf) { out_idx_buf[pos] = in_idx_buf ? in_idx_buf[i] : i; }
+        int bucket = calc_bucket<T, BitsPerPass>(value, start_bit, mask, greater);
+        atomicAdd(histogram_smem + bucket, IdxT(1));
+
+        if (counter_len == 1) {
+          out[k - 1]     = value;
+          out_idx[k - 1] = in_idx_buf ? in_idx_buf[i] : i;
+        }
+      } else if (out && prev_bucket < want_bucket) {
```

Comment:
Is out always a valid pointer? Three lines above we do not check, but here we do. Please make it consistent.